[3.3] Additional identifiers using extended data importer

Atticus29 opened the pull request:
# Description

This PR implements bulk upload of custom identifiers. End users need to provide either occurrenceID, catalogNumber, or otherCatalogNumbers, as well as identifierName and identifierValue. format, notes, sortBy, and recordID are optional.

End users also have the option to update existing identifier entries in the omoccuridentifiers table, as well as delete.

# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [ ] Hotfixes should be branched off of the `master` branch and PR'd using the **merge** option (not squashed) into the `hotfix` branch.
    - N/A
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [x] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
- [x] There are no linter errors
- [x] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
- [x] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [ ] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
    - Autoformats are in their own commits, and comments are in places where autoformatters affected the whole file but certain code is new.
- [x] Comment which GitHub issue(s), if any does this PR address
    - [https://github.com/orgs/BioKIC/projects/6/views/32?pane=issue&itemId=89420848](https://github.com/orgs/BioKIC/projects/6/views/32?pane=issue&itemId=89420848)
- [ ] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).
    - N/A

# Post-Approval

- [ ] It is the code author's responsibility to merge their own pull request after it has been approved
- [ ] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [ ] If this PR represents a merge into the `hotfix` branch, remember to use the **merge** option (i.e., no squash).
- [ ] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [ ] If this PR represents a merge from the `hotfix` branch into the `master` branch use the **squash & merge** option
  - [ ] a subsequent PR from `master` into `Development` should be made with the **merge** option (i.e., no squash).
  - [ ] **Immediately** delete the `hotfix` branch and create a new `hotfix` branch
  - [ ] increment the Symbiota version number in the symbase.php file and commit to the `hotfix` branch.
- [ ] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [ ] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!

Atticus29 committed:
add translations

Atticus29 committed:
autoformat importextended, improve readability of importType option elements and add new option for identifiers

Atticus29 committed:
add more translations

Atticus29 committed:
begin to implement OmIdentifiers.php class

Atticus29 committed:
add logic to handle selection of identifier for export and autoformat

Atticus29 committed:
add Identifier stuff to setTargetFieldArr

Atticus29 committed:
fix errors preventing the run of insertIdentifier

Atticus29 committed:
get import to technically work

Atticus29 committed:
get update import to technically work

Atticus29 committed:
remove initialTimestamp, modifiedTimestamp, and modifiedUid from the upload file and address generating those data on the fly. Fixed for insertIdentifier but still out of order for updateIdentifier

Atticus29 committed:
fix update order

Atticus29 committed:
remove occid as an option and add catalogNumber and occurrenceID

Atticus29 committed:
get delete to work

Atticus29 committed:
remove cruft

Atticus29 committed:
resolve all but one merge conflict

Atticus29 committed:
finish resolving merge conflict in collections/admin/importextended.php hopefully

Atticus29 committed:
remove more cruft

Atticus29 commented on classes/OccurrenceImport.php:
```diff
@@ -160,287 +166,378 @@ private function insertRecord($recordArr, $occidArr, $postArr){
+					$this->logOrEcho('ERROR loading Material Sample: ' . $importManager->getErrorMessage(), 1);
+				}
+			}
+		} elseif ($this->importType == self::IMPORT_IDENTIFIERS) {
```
This is the only area that is a new addition; the rest is autoformatting fixes.

themerekat commented:
![image](https://github.com/user-attachments/assets/5ac3548b-5cf1-466b-be1a-5391dbfecbb7)
What I did:

1. Uploaded a file of identifiers. One of the identifiers matched on otherCatalogNumbers, and, incidentally, that same otherCatalogNumber has TWO matching occurrences already in the database. So what it did the first time was add the identifier to both instances of that otherCatalogNumber.
2. Changed one value in the identifiers field and re-uploaded the file. I just realized that I failed to check the "update" box, though, so it's just trying to add brand new identifier records to the db. It's weird that it worked on the # 1, but failed on # 2, but I suspect it has to do with the fact that # 2 has two matching occurrence records in the db already.

Atticus29 committed:
fix bug where typeStr was not being reinitialized after each sql call

Atticus29 commented:
> ![image](https://private-user-images.githubusercontent.com/16688784/410231396-5ac3548b-5cf1-466b-be1a-5391dbfecbb7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NzUxMDIsIm5iZiI6MTczODg3NDgwMiwicGF0aCI6Ii8xNjY4ODc4NC80MTAyMzEzOTYtNWFjMzU0OGItNWNmMS00NjZiLWJlMWEtNTM5MWRiZmVjYmI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDIwNDY0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU0YTdhYjVlNDc0ZjJkZTE3MGM2ODVmMDZlODBhMTllMjdmMjFhMGVmNTA4YjJkMGU2MTUxMWQyZDQxNTI4NGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0EiTzuMcV801H8c7U_OeeGQ3T39-HuA-Y_iTYt3HoDY) What I did:
> 
> 1. Uploaded a file of identifiers. One of the identifiers matched on otherCatalogNumbers, and, incidentally, that same otherCatalogNumber has TWO matching occurrences already in the database. So what it did the first time was add the identifier to both instances of that otherCatalogNumber.
> 2. Changed one value in the identifiers field and re-uploaded the file. I just realized that I failed to check the "update" box, though, so it's just trying to add brand new identifier records to the db. It's weird that it worked on the # 1, but failed on # 2, but I suspect it has to do with the fact that # 2 has two matching occurrence records in the db already.

Ok. I think that I fixed this in 578a7a5c9, but @themerekat it's worth checking that it has the result that you're expecting.

themerekat commented:
The DELETE action doesn't seem to work. Here's the error I'm getting:
![image](https://github.com/user-attachments/assets/3af4ae94-a88a-4e41-a6a8-69e929273f63)

Oh wait...I didn't map "identifiername" or "identifiervalue", just the subject identifier.

themerekat reviewed (approved):
Add, update, and delete all seem to be working!

Atticus29 committed:
Merge branch 'Development' into additional-identifiers-using-extended-data-importer

Atticus29 committed:
resolve merge conflicts in collections/datasets/datasetmanager.php

egbot committed:
Minor adjustments standardizing code

- Remove auto-format that put starting bracket on new line
- Changed 4-spaces to tabs
- Rearranged order of functions within OmIdentifiers to match other Om data files

egbot committed:
Temporary format reversal

egbot committed:
Merge branch 'additional-identifiers-using-extended-data-importer' into temp-resolve-conflict-identifier-import

egbot committed:
Merge pull request #2344 from Symbiota/temp-resolve-conflict-identifier-import

Update branch with latest Development branch modifications

egbot committed:
Reapply previous autoformat

Atticus29 committed:
fix issue where missing identifier name or identifier value were not alerting

Atticus29 commented:
> If "subject identifier" is not mapped, js is used display a message that this is a requiered field, which is great. However, the other required fields (identifierName and identierValue) don't stop the form from being submitted and displays errors for all records due to missing fields. Can this also be validated prior to form submit?

Fixed in a17649575

Atticus29 committed:
fix logic to communicate updates or deletions in addition to additions

Atticus29 commented:
> When an identifier is replaced (via match on name and value), the message displayed is that it was added. Instead, can it state that it was "replaced"? And if it is deleted, can it state that it was deleted?

Fixed in fbba46527

Atticus29 committed:
remove recordID and comment out format, notes, and sortBy from the upload columns fieldArr

Atticus29 commented:
> recordID should be removed from the select list. The recordID fields are only meant to be populated internally with a UUID and never changed. Users are never allowed to add or change this field.
> I like the idea of allowing notes, format, and sortBy to be populated and changed via this interface. However, since they can't yet be viewed or changed within the editor, they should probably be commented out for now. We should probably create another PR to add these fields to the editor interface, and they these can be included in the import.

Fixed in a1ffc8e45

Atticus29 committed:
add occid as an identifier by which updates and adds can happen

Atticus29 commented:
> Is occid meant to be a method for matching on source record identifier (subject identifier)? If so, it should be added to the above "subject identifier" list.
Fixed in 60e55f83e.

Atticus29 commented:
> Are there instructions within the Doc yet? If so, we should include a link with the other instruction links. Or perhaps this is to be a secondary PR? By the way, I previously added the instruction links as a bullet list. Feel free to change this if you can think of a better way to display the links.
@themerekat maybe we can add this to our documentation discussion tomorrow?

themerekat commented:
> > Are there instructions within the Doc yet? If so, we should include a link with the other instruction links. Or perhaps this is to be a secondary PR? By the way, I previously added the instruction links as a bullet list. Feel free to change this if you can think of a better way to display the links.
> > @themerekat maybe we can add this to our documentation discussion tomorrow?

We haven't drafted docs yet! I think it's a great idea to add to our discussion today.

Atticus29 commented:
> > > Are there instructions within the Doc yet? If so, we should include a link with the other instruction links. Or perhaps this is to be a secondary PR? By the way, I previously added the instruction links as a bullet list. Feel free to change this if you can think of a better way to display the links.
> > > @themerekat maybe we can add this to our documentation discussion tomorrow?
> 
> We haven't drafted docs yet! I think it's a great idea to add to our discussion today.

Issue created:
[https://github.com/Symbiota/Symbiota-Documentation/issues/5](https://github.com/Symbiota/Symbiota-Documentation/issues/5).
